### PR TITLE
[SPARK-23909][SQL] Add filter function.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

This pr adds `filter` function which filters the input array using the given predicate.

```sql
> SELECT filter(array(1, 2, 3), x -> x % 2 == 1);
 array(1, 3)
```

## How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by ueshin** [comment]
This is based on #21954. I'll update after it is merged.

**Comment by gatorsmile** [comment]
@ueshin Please rebase it. Thanks!

**Comment by gatorsmile** [comment]
cc @hvanhovell

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by kiszk** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala [hunk 1]
```diff
@@ -210,3 +219,54 @@ case class ArrayTransform(
...
+
+  @transient lazy val LambdaFunction(_, Seq(elementVar: NamedLambdaVariable), _) = function
+
+  override def eval(input: InternalRow): Any = {
+    val arr = this.input.eval(input).asInstanceOf[ArrayData]
+    if (arr == null) {
+      null
+    } else {
+      val f = functionForEval
+      val buffer = new mutable.ArrayBuffer[Any]
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala]: I am wondering if we should use the buffer builder with a size hint here? Or, alternatively manage the array ourself.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala [hunk 2]
```diff
@@ -210,3 +219,54 @@ case class ArrayTransform(
...
+        if (f.eval(input).asInstanceOf[Boolean]) {
+          buffer += elementVar.value.get
+        }
+        i += 1
+      }
+      new GenericArrayData(buffer)
+    }
+  }
+
+  override def prettyName: String = "filter"
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala]: Is filter too generic? wdyt?

**Review by hvanhovell — APPROVED** [review]
LGTM - two small nits, that can be addressed later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala (hunk 1 above)
Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala]: Sounds good. I'll use the original length as a size hint.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala (hunk 2 above)
Comment (ueshin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala]: Yeah, it might be. How about `array_filter`?

**Comment by kiszk** [comment]
retest this please

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**